AWS INIT Documentations (#3661)
wso2/docs-apim: en/docs/deploy-and-publish/publish-on-dev-portal/publish-aws-apis-in-the-dev-portal.md

# Importing APIs from AWS API Gateway to the WSO2 API Manager  

> This feature helps publish an API a user has in the AWS API gateway, on the WSO2 developer portal. These APIs will not be deployed into the WSO2 API gateway and will not be managed by the WSO2 API gateway.

Reviewer: ```suggestion
This feature helps publish an API a user has in the AWS API gateway, on the WSO2 Developer portal. These APIs will not be deployed into the WSO2 API gateway and will not be managed by the WSO2 API gateway.
```

Reviewer: Please check the other places as well

## Prerequisites 

> 1. Two WSO2 API Controller commands are needed for this task. Hence the user needs to have the APICTL downloaded and configured. The apictl is a command line interface similar to the AWS CLI. [Getting started with the API Controller](https://apim.docs.wso2.com/en/latest/learn/api-controller/getting-started-with-wso2-api-controller/)

Reviewer: ```suggestion
1. Two WSO2 API Controller (apictl) commands are needed for this task. Hence the user needs to have the apictl downloaded and configured. The apictl is a command line interface similar to the AWS CLI. [Getting started with the API Controller](https://apim.docs.wso2.com/en/latest/learn/api-controller/getting-started-with-wso2-api-controller/)
```

Reviewer: IMO, the sentence "The apictl is a command line interface similar to the AWS CLI" makes apictl alien to WSO2 API-M. The WSO2 product here is apictl. So I do not think it is suitable to say that "apictl is similar to AWS CLI" since AWS CLI is not the WSO2 product.

Reviewer 2: Agreed, we should directly compare apictl with AWS CLI, specially since this is a feature that uses the AWS CLI via the apictl, this can sound confusing. Lets just drop the sentence `The apictl is a command line interface similar to the AWS CLI`

Author: should i completely drop that sentence or change it to "the apictl is a command line interface" ?

Reviewer: Shall we add something like this after dropping the part which you mentioned AWS CLI as @uvindra stated?

```suggestion
1. Two (2) **WSO2 API Controller (apictl)** commands are needed for this task. Hence the user needs to have the apictl downloaded and configured. (apictl is a command-line tool providing the capability to move APIs, API Products, and Applications across environments and to perform CI/CD operations. For more information, see [Getting started with the API Controller](https://apim.docs.wso2.com/en/latest/learn/api-controller/getting-started-with-wso2-api-controller/))
```

Reviewer: Another tip, shall we cross-check all the links in this doc? Especially, the above link you mentioned points to the **latest** doc space which is 3.2.0. We should always have the links specified by the versions. In your case it should be **4.0.0**. Let's check if there are any occurrences like this and correct those to point to the correct doc space page (4.0.0).

Reviewer: The best way to mention the links is with the base path, so that we do not need to worry about the version. 

Example:- 
[Getting Started with WSO2 API Controller (apictl)]({{base_path}}/install-and-setup/setup/api-controller/getting-started-with-wso2-api-controller)

> 1. Two WSO2 API Controller commands are needed for this task. Hence the user needs to have the APICTL downloaded and configured. The apictl is a command line interface similar to the AWS CLI. [Getting started with the API Controller](https://apim.docs.wso2.com/en/latest/learn/api-controller/getting-started-with-wso2-api-controller/)

Reviewer 3: Shall we change this as follows?
```suggestion
1. Two WSO2 API Controller commands are needed for this task. Hence the user needs to have the APICTL downloaded and configured. The apictl is a command-line interface similar to the AWS CLI. [Getting started with the API Controller](https://apim.docs.wso2.com/en/latest/learn/api-controller/getting-started-with-wso2-api-controller/)
```

Reviewer 2: As mentioned previously, lets drop the sentence `The apictl is a command line interface similar to the AWS CLI`

2. The `aws init` command will be using the AWS CLI in the background. Hence the user needs to download and install the AWS CLI before using the `aws init` command. Refer [how to download and install the AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html).

3. After successfully installing the AWS CLI it needs to be configured. Refer [how to configure the AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-configure.html).

> 4. Enable AWS CLI pagination. By default the AWS CLI returns all objects in the output, but if you have disabled pagination you will have to enable pagination for the `aws init` command to successfully work. 

Reviewer 3: Shall we add the comma here?
```suggestion
4. Enable AWS CLI pagination. By default, the AWS CLI returns all objects in the output, but if you have disabled pagination you will have to enable pagination for the `aws init` command to successfully work. 
```

For more information, refer [here](https://docs.aws.amazon.com/cli/latest/userguide/cli-usage-pagination.html).

> 5. The API Manager server needs to be up and running to import the API project to the APIM successfully.

Reviewer: ```suggestion
5. The WSO2 API Manager (WSO2 API-M) server needs to be up and running to import the API project to the APIM successfully.
```

6. An active internet connection is also required to download the OAS of the AWS API to initialize the API project.

## Importing an API from AWS API gateway to the WSO2 API Manager  

Usage of the `aws init` command is quite simple. Two mandatory flags are associated with the `aws init` command, `name` and `stage`. `name` will be the name of the API that is in the AWS API gateway that the user wants to import and publish in the developer portal and `stage` will be the specific stage of that API.

-   **Command**

    ``` bash
    apictl aws init -n <aws-api-name> -s <api-stage-to-get>
    ```

!!! example

    ```bash
    apictl aws init -n PetStore -s demo
    ```

!!! info
    **Flags:**  
            
    -    Required :  

        `--name` or `-n` : Name of the API to get from the AWS API Gateway

> 

Reviewer: Shall we remove the extra line here?

        `--stage` or `-s` : API stage to get 

    -   Optional :

>         `-f` : To force create an API project

Reviewer 3: Don't we have the `--force` flag supported with this command like the other commands in apictl?

Eg: https://github.com/wso2/product-apim-tooling/blob/b14f927c4d9100a78932abe34fc4b79a29ea5df3/import-export-cli/cmd/init.go#L112

Author: we do @Chamindu36. `-f` is the shortened flag right ?

Reviewer 3: Yes -f is used as the shortened version.

> 
>         `--name` or `-n` : Name of the API to get from the AWS API Gateway
> 
>         `--stage` or `-s` : API stage to get 
> 
>     -   Optional :
> 
>         `-f` : To force create an API project
> 
>         `--verbose` or `-v` : To execute the command in verbose mode

Reviewer: ```suggestion
        `--name` or `-n` : Name of the API to get from the AWS API Gateway
        `--stage` or `-s` : API stage to get 
    -   Optional :
        `-f` : To force create an API project
        `--verbose` or `-v` : To execute the command in verbose mode
```

Reviewer: Let's remove all the extra lines here. (You can refer to other API Controller doc pages to get an idea of how we have mentioned Required and Optional flags of apictl commands)

!!! tip

>     The API Manager project will be initialized in the same location where the command was executed from. 

Reviewer: ```suggestion
    The API project will be initialized in the same location where the command was executed from. 
```

-   **Response**
    
    ``` bash tab="Response Format"
    Initializing a new WSO2 API Manager project in /<path-to-api-manager-project>
    Project initialized
    Open README file to learn more
    ```
    
    ``` bash tab="Example Response"
    Initializing a new WSO2 API Manager project in /home/user/Documents/product-apim-tooling/import-export-cli/products
    Project initialized
    Open README file to learn more
    ```
### How the **aws init** command works.

The `aws init` command will be using the AWS CLI in the background to get a list of all the APIs from the AWS API gateway. It will then look for the ID of the API name provided by the user when the `aws init` command was executed, to download the OAS of that API. After making the necessary adjustments to the APIs version, tags, endpoints etc, an API project will be initialized and the downloaded OAS will be deleted.

### Importing the initialized API project to the WSO2 API Manager as an API

> Upon successful initialization of the project, execute the following command to import that API project to the WSO2 publisher portal as an API. 

Reviewer: ```suggestion
Upon successful initialization of the project, execute the following command to import that API project to the WSO2 Publisher portal as an API. 
```

Reviewer: Please check the other places as well

> Upon successful initialization of the project, execute the following command to import that API project to the WSO2 publisher portal as an API. 

Reviewer 3: ```suggestion
Upon successful initialization of the project, execute the following command to import that API project to the WSO2 API-M publisher portal as an API. 
```

    ```

!!! example
    ```bash
    apictl import api -f /home/user/Documents/product-apim-tooling/import-export-cli/products -e production -k
    ```

After successfully importing the API, it will appear in the publisher as shown below.

[![]({{base_path}}/assets/img/publish/aws-api-publisher.png)]({{base_path}}/assets/img/publish/aws-api-publisher.png)

To publish the API, go to `Lifecycle` under `Publish` and click on `Publish`.

[![]({{base_path}}/assets/img/publish/publish-aws-api.png)]({{base_path}}/assets/img/publish/publish-aws-api.png)

To learn more about importing an API project refer to [import-api-project](https://apim.docs.wso2.com/en/latest/learn/api-controller/importing-apis-via-dev-first-approach/#import-an-api-project).

The initial state of the API will be set to `CREATED` so the user will be able to make changes to the API if necessary in the publisher portal, before publishing the API on the developer portal.

## Troubleshooting errors

If any errors are being printed and if the problem is not clear, execute the `aws init` command with the `--verbose` flag to get a more detailed output.

### Unable to find an API with the name `<api-name>`

> After making sure the user has provided the correct API name and that the API is in the AWS API gateway, if the problem persists, it is possible that the user has disabled AWS CLI pagination which will cause the AWS CLI to output only a limited number of APIs. In such cases, the API the user is looking for might not be in the APIs list provided by the AWS CLI. Enable AWS CLI pagination and try again. 

Reviewer 3: Shall we add this small change?
```suggestion
After making sure the user has provided the correct API name and that the API is in the AWS API gateway if the problem persists, it is possible that the user has disabled AWS CLI pagination which will cause the AWS CLI to output only a limited number of APIs. In such cases, the API the user is looking for might not be in the APIs list provided by the AWS CLI. Enable AWS CLI pagination and try again. 
```

> After making sure the user has provided the correct API name and that the API is in the AWS API gateway, if the problem persists, it is possible that the user has disabled AWS CLI pagination which will cause the AWS CLI to output only a limited number of APIs. In such cases, the API the user is looking for might not be in the APIs list provided by the AWS CLI. Enable AWS CLI pagination and try again. 

Reviewer 4: How about we link the pagination docs again in here?

### Error code 255

One of the most common errors related to this command is the `exit status 255`. This is a AWS CLI error code and it means the AWS CLI command failed and that there were errors generated by the AWS CLI or by the AWS service to which the request was sent.
The most common reason for this is due to a incorrect stage name provided.

!!! tip
    Make sure you provide the correct stage name and the spellings are correct.

>     It is also important to remember the name and stagename are case sensitive.

Reviewer 3: ```suggestion
    It is also important to remember the `name` and `stage`. `name` are case sensitive.
```

Author: @Chamindu36 `name` and `stagename` are two different things. stage-name is the stage of that API.

Reviewer 3: Yes. Isn't it good if we can use the text style (the style used in commands and flags) as we do in the other places in the documentation for these properties without using the normal style

Author: yes @Chamindu36 will do that. but i meant the second change you had mentioned ( `stage`, `name`). That is not required right ? because their `stagename` is one.

Reviewer 3: Yes. We can omit that
